repa/grids/psdiffusion.cpp

```diff
@@ -0,0 +1,158 @@
+ *
+ * Repa is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Repa.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "psdiffusion.hpp"
+
+#include <algorithm>
+#include <set>
+
+#include "grids/util/mpi_cart.hpp"
+#include "grids/util/mpi_graph.hpp"
+#include "util/vec_arith.hpp"
+
+#ifndef NDEBUG
+#define PSDIFFUSION_DEBUG
+#endif
+
+namespace repa {
+namespace grids {
+
+static Vec3i
+fix_periodic_edge(const Vec3i &c0, const Vec3i &c2, const Vec3i &gsize)
+{
+    using namespace util::vector_arithmetic;
+    Vec3i te, ts;
+    for (int i = 0; i < te.size(); i++) {
+        te[i] = c2[i] - c0[i] > 1;
+        ts[i] = c0[i] - c2[i] > 1;
+    }
+    return c2 + ts * gsize - te * gsize;
+}
+
+/*
+ * Initialization
+ */
+PSDiffusion::PSDiffusion(const boost::mpi::communicator &comm,
+                         Vec3d box_size,
+                         double min_cell_size)
+    : Diffusion(comm, box_size, min_cell_size)
+{
+}
+
+PSDiffusion::~PSDiffusion()
+{
+}
+
+void PSDiffusion::post_init(bool firstcall)
+{
+    Diffusion::post_init(firstcall);
+    comm_dims = util::mpi_cart_get_dims(comm_cart);
+
+#ifdef PSDIFFUSION_DEBUG
+    if (firstcall) {
+        initial_neighborhood
+            = std::vector<rank_type>(neighbors.begin(), neighbors.end());
+
+        // Allgather over neighbors.size
+        std::vector<int> neighbors_sizes(neighbors.size());
+        int nsize = neighbors.size();
+        MPI_Neighbor_allgather(&nsize, 1, MPI_INT, neighbors_sizes.data(), 1,
+                               MPI_INT, neighcomm);
+        // Assert neighbors.size equal
+        assert(std::equal(neighbors_sizes.begin() + 1, neighbors_sizes.end(),
+                          neighbors_sizes.begin()));
+
+        neighborhood_ranks.resize(neighbors.size() * neighbors.size());
+        MPI_Neighbor_allgather(neighbors.data(), neighbors.size(), MPI_INT,
+                               neighborhood_ranks.data(), neighbors.size(),
+                               MPI_INT, neighcomm);
+
+        for (int i = 0; i < neighbors.size(); i++)
+            nr_mappings[neighbors[i]]
+                = &neighborhood_ranks.data()[0] + (neighbors.size() * i);
+    }
+
+    // Neighborhood consistence check
+    std::set<rank_type> a(initial_neighborhood.begin(),
+                          initial_neighborhood.end());
+    std::set<rank_type> b(neighbors.begin(), neighbors.end());
+
+    assert(std::includes(a.begin(), a.end(), b.begin(), b.end()));
+#endif
+}
+
+bool PSDiffusion::accept_transfer(local_cell_index_type cidx,
+                                  rank_type neighrank) const
+{
+    const bool b1 = coords_based_allow_sending(cidx, neighrank);
+#ifdef PSDIFFUSION_DEBUG
```

Comment:
For the merged version, please completely remove the `rank_based_allow_sending` function (as it is the slower one, right?).

Reply:
Yes it is the slower version. Removed in 14f7a25